src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
+                    }
+                    else
+                    {
+                        // We initiated a connection creation.  When it completes, transfer the result to the waiter.
+                        connectionTask.AsTask().ContinueWith((innerConnectionTask, state) =>
+                        {
+                            var innerWaiter = (ConnectionWaiter)state;
+                            try
+                            {
+                                // Get the resulting connection.
+                                HttpConnection result = innerConnectionTask.GetAwaiter().GetResult();
+
+                                // Store the resulting connection into the waiter. As in the synchronous case,
+                                // since we already have a count that's inflated due to the connection being
+                                // disassociated, we don't need to change the count here.
+                                innerWaiter.SetResult(innerConnectionTask.Result);
+                            }
+                            catch (Exception e)
+                            {
+                                // The creation operation failed.  Store the exception into the waiter.
+                                innerWaiter.SetException(e);
 
-                foreach (HttpConnection connection in _idleConnections)
+                                // At this point, a connection was dropped and we failed to replace it,
+                                // which means our connection count still needs to be decremented.
+                                innerWaiter._pool.DecrementConnectionCount();
+                            }
+                        }, waiter, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+                    }
+                }
+            }
+        }
+        
+        /// <summary>Returns the connection to the pool for subsequent reuse.</summary>
+        /// <param name="connection">The connection to return.</param>
+        public void ReturnConnection(HttpConnection connection)
+        {
+            List<CachedConnection> list = _idleConnections;
+            lock (SyncObj)
+            {
+                Debug.Assert(list.Count <= _maxConnections, $"Expected {list.Count} <= {_maxConnections}");
+
+                // Mark the pool as still being active.
+                _usedSinceLastCleanup = true;
+
+                // If there's someone waiting for a connection, simply
+                // transfer this one to them rather than pooling it.
+                if (_waiters != null && _waiters.TryDequeue(out ConnectionWaiter waiter))
+                {
+                    waiter.SetResult(connection);
+                    return;
+                }
+
+                // If the pool has been disposed of, dispose the connection
+                // being returned, as the pool is being deactivated.
+                if (_disposed)
```

Comment:
> Any reason we do it the way we do?

Yes.  Imagine you concurrently schedule 100 requests to a pool with a 50 limit, and then immediately after creating them you dispose of the pool.  Those 100 requests should still all be processed, and so you want the latter 50 to inherit the connections of the former 50. You only want those connections disposed of if the pool is disposed of and there's no one waiting to transfer the connections to.